src/ray/core_worker/core_worker.h

```diff
@@ -0,0 +1,62 @@
+#ifndef RAY_CORE_WORKER_CORE_WORKER_H
+#define RAY_CORE_WORKER_CORE_WORKER_H
+
+#include "common.h"
+#include "object_interface.h"
+#include "ray/common/buffer.h"
+#include "task_execution.h"
+#include "task_interface.h"
+
+namespace ray {
+
+/// The root class that contains all the core and language-independent functionalities
+/// of the worker. This class is supposed to be used to implement app-language (Java,
+/// Python, etc) workers.
+class CoreWorker {
+ public:
+  /// Construct a CoreWorker instance.
+  ///
+  /// \param[in] worker_type Type of this worker.
+  /// \param[in] langauge Language of this worker.
+  CoreWorker(const WorkerType worker_type, const Language language)
+      : worker_type_(worker_type),
+        language_(language),
+        task_interface_(this),
+        object_interface_(this),
+        task_execution_interface_(this){};
```

Comment:
may you explain a bit on why various interfaces need to have a back pointer to the `CoreWorker` class? Usually we use shared_ptr to hold a reference but here CoreWorker shouldn't destructed as long as the worker is alive. Or do we expect various interfaces to potentially call each other using this pointer?

Reply:
btw. if the back shared_ptr is indeed required, this class needs to be derived from `enable_shared_from_this` ?

Reply:
Good point, if necessary, the reason should be documented.

Reply:
Thanks, I tried using `enable_shared_from_this` but it turned out that we can't use `shared_from_this` in the constructor. 
I'll change it back to normal pointers, because, as you mentioned, lifecycle of the sub instances is the same as the root CoreWorker instance. cc @pcmoritz 

Also, the reason why we need a back pointer is because we would need to use object store methods in the task interface.

Reply:
thanks. For the back pointer, you can instead use reference, which should also work here and also ensures validity by design.

For why back pointer is needed, yesterday I prototyped the task submission changes, and found that things like WorkerContext would also be needed by multiple interfaces (submission, execution, and probably also object store).